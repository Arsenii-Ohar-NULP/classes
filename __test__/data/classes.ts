import Class from 'components/classes/Class';

export const sampleFiveClasses: Class[] = [
    {
        id: 1,
        title: 'Class 1',
        description: 'This is class 1',
        teacher_id: 1,
        students_number: 1
    },
    {
        id: 2,
        title: 'Class 2',
        description: 'This is class 2',
        teacher_id: 2,
        students_number: 1
    },
    {
        id: 3,
        title: 'Class 3',
        description: 'This is class 3',
        teacher_id: 3,
        students_number: 1

    },
    {
        id: 4,
        title: 'Class 4',
        description: 'This is class 4',
        teacher_id: 4,
        students_number: 1
    },
    {
        id: 5,
        title: 'Class 5',
        description: 'This is class 5',
        teacher_id: 5,
        students_number: 1
    }
]