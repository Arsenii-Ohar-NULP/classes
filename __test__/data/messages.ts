import Message from 'pages/class/Message';

const sampleMessages : Message[] = [
    {
        id: 1,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 2,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 3,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    }

]