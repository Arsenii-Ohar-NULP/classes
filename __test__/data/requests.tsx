import { JoinRequest } from 'pages/class/JoinRequest'

export const getRequestsWithClassId = (classId: number, number: number) => {
    const requests: JoinRequest[] = []
    for (let i = 0; i < number; i++){
        requests.push(
            {
                userId: i + 1,
                classId: classId
            }
        )
    }

    return requests;
}