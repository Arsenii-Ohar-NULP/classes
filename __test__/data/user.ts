import User, { Role } from 'components/account/User';

export const sampleUser: User = {
    id: 1,
    username: 'seniorohar',
    firstName: 'Arsenii',
    lastName: 'Ohar',
    email: 'arsen.ogar@gmail.com',
    phone: '+380987669293',
    role: Role.Teacher   
}

export const sampleFiveUsers: User[] = [
    {
        id: 1,
        username: 'seniorohar',
        firstName: 'Arsenii',
        lastName: 'Ohar',
        email: 'arsen.ogar@gmail.com',
        phone: '+380987669293',
        role: Role.Student
    },
    {
        id: 2,
    username: 'vitalikpikh',
    firstName: 'Vitalik',
    lastName: 'Pikhotskiy',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
    {
        id: 3,
    username: 'svyatoslavshainoha',
    firstName: 'Svyatoslav',
    lastName: 'Shainoha',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
      {
        id: 4,
    username: 'olehrysin',
    firstName: 'Oleh',
    lastName: 'Rysin',
    email: 'olehrysin@gmail.com',
    phone: '+380967323273',
    role: Role.Student
    },
    {
        id: 5,
    username: 'johnadams',
    firstName: 'John',
    lastName: 'Adams',
    email: 'johnadams@gmail.com',
    phone: '+380967723273',
    role: Role.Student
    }
]