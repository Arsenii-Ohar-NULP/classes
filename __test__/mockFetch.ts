export const mockFetch = () => {
    global.fetch = jest.fn();
}