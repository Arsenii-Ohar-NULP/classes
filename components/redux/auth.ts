import { createSlice } from '@reduxjs/toolkit';
import User from 'components/account/User';

export enum AuthStatus {
  LOGGED_IN_FETCHED,
  LOGGED_IN,
  LOGGED_OUT
}

export interface IAuthState {
  status: AuthStatus;
  user: User;
}

const initialState: IAuthState = { status: AuthStatus.LOGGED_OUT, user: null };

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    fetch(state){
      state.status = AuthStatus.LOGGED_IN_FETCHED;
    },
    login(state, action) {
      return {
        ...state,
        status: AuthStatus.LOGGED_IN,
        user: action.payload,
      };
    },
    logout(state) {
      return {
        ...state,
        status: AuthStatus.LOGGED_OUT,
        user: null,
      };
    },

    updateUser(state, action) {
      return {
        ...state,
        user: { ...state.user, ...action.payload },
      };
    },
  },
});

export const authInitialState = initialState; 
export const authActions = authSlice.actions;
