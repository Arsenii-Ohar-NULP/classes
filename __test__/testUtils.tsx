import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
