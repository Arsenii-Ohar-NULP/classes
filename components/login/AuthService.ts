import {useRouter} from 'next/navigation';
import {authActions} from 'components/redux/auth';
import {useAppDispatch} from 'components/redux/store';

export const accessTokenKey = 'accessToken';

export interface Credentials {
  username: string;
  password: string;
}

export const logout = (dispatch, router) => {
  removeToken();
  dispatch(authActions.logout());
  router.replace('/auth/login');
};

export const useLogout = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();

  return () => logout(dispatch, router);
};

const saveInSession = (token: string) => {
  sessionStorage.setItem(accessTokenKey, token);
};

const saveInLocal = (token: string) => {
  localStorage.setItem(accessTokenKey, token);
};

export const saveToken = (token: string, remember: boolean) => {
  if (!remember) {
    saveInSession(token);
    return;
  }

  saveInLocal(token);
};

export const removeToken = () => {
  if (sessionStorage.getItem(accessTokenKey)) {
    sessionStorage.removeItem(accessTokenKey);
  }

  if (localStorage.getItem(accessTokenKey)) {
    localStorage.removeItem(accessTokenKey);
  }
};
