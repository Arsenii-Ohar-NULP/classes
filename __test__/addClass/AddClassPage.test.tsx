import * as React from 'react';

import {renderWithProviders} from '__test__/testUtils';
import AddClassPage from 'app/main/classes/addClass/page';
import {sampleUser} from '__test__/data/user';
import {AuthStatus} from 'components/redux/auth';
import {fireEvent, screen, waitFor} from '@testing-library/react';
import {useLogout} from 'components/login/AuthService';
import {server} from "../api/server";
import {rest} from "msw";
import {CLASSES_API_URL} from "../../components/redux/utils";
import {useRouter} from "next/navigation";


const pushMock = jest.fn();
jest.mock('next/navigation', () => ({
    useRouter: () => ({
        push: pushMock,
        replace: pushMock
    })
}))

jest.mock('components/login/AuthService')

describe('add class page', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    })

    it('should match a snapshot', () => {
        const page = renderWithProviders(<AddClassPage/>, {
            preloadedState: {
                auth: {
                    status: AuthStatus.LOGGED_IN_FETCHED,
                    user: sampleUser
                },
                classes: {
                    userClasses: [],
                    joinRequests: []
                },
                search: {
                    students: null,
                    classes: null
                }
            }
        })

        expect(page.container).toMatchSnapshot();
    })

    it('should navigate to the main page, when user typed in valid data and clicked the button', async () => {
        server.use(rest.post(`${CLASSES_API_URL}/class`, (req, res, clx) =>
            res(clx.status(200), clx.json(200))
        ))
        server.use(rest.post(`${CLASSES_API_URL}/class/img`, (req, res, ctx) =>
            res(ctx.status(201))
        ))
        const user = sampleUser;
        const testFile = new File(['hello'], 'hello.png', {type: 'image/png'})
        renderWithProviders(<AddClassPage/>, {
            preloadedState: {
                auth: {
                    status: AuthStatus.LOGGED_IN_FETCHED,
                    user
                },
                classes: {
                    userClasses: [],
                    joinRequests: []
                },
                search: {
                    students: null,
                    classes: null
                }
            }
        })
        const titleInput = screen.getByPlaceholderText('Enter a title');
        fireEvent.change(titleInput, {target: {value: 'React.js in a year'}})

        const descriptionInput = screen.getByPlaceholderText('Enter a description');
        fireEvent.change(descriptionInput, {target: {value: 'Here you can learn React.js in a year'}});

        const fileInput = screen.getByTestId('file-input');
        fireEvent.change(fileInput, {target: {files: [testFile]}});

        const createButton = screen.getByText('Add');
        fireEvent.click(createButton);

        await waitFor(() => {
            expect(useRouter().replace).toHaveBeenCalledWith("/main/classes");
        })
    })

    it('should logout a user, when one of the API requests returns 401', async () => {
        server.use(rest.post(`${CLASSES_API_URL}/class`, (req, res, clx) =>
            res(clx.status(401))
        ))
        const user = sampleUser;
        const testFile = new File(['hello'], 'hello.png', {type: 'image/png'})
        const mockedLogout = jest.fn();
        jest.mocked(useLogout).mockReturnValue(mockedLogout);
        renderWithProviders(<AddClassPage/>, {
            preloadedState: {
                auth: {
                    status: AuthStatus.LOGGED_IN_FETCHED,
                    user
                },
                classes: {
                    userClasses: [],
                    joinRequests: []
                },
                search: {
                    students: null,
                    classes: null
                }
            }
        })
        const titleInput = screen.getByPlaceholderText('Enter a title');
        fireEvent.change(titleInput, {target: {value: 'React.js in a year'}})

        const descriptionInput = screen.getByPlaceholderText('Enter a description');
        fireEvent.change(descriptionInput, {target: {value: 'Here you can learn React.js in a year'}});

        const fileInput = screen.getByTestId('file-input');
        fireEvent.change(fileInput, {target: {files: [testFile]}});

        const createButton = screen.getByText('Add');
        fireEvent.click(createButton);

        await waitFor(() => {
            expect(mockedLogout).toHaveBeenCalled();
        })

    })
})