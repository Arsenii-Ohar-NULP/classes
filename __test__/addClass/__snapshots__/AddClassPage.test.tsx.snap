// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add class page should match a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="container p-1 d-flex flex-column align-items-center"
      >
        <h1
          class="p-2"
        >
          Add a class
        </h1>
        <input
          class="form-control rounded p-1 px-2"
          id="title"
          name="Title"
          placeholder="Enter a title"
        />
        <div
          class="invalid"
        />
        <textarea
          class="form-control rounded p-1 px-2"
          id="description"
          name="Description"
          placeholder="Enter a description"
        />
        <div
          class="invalid"
        />
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="formFile"
          >
            Upload a class thumbnail
          </label>
          <input
            accept=".jpg, .png"
            class="form-control form-control-sm"
            data-testid="file-input"
            id="ImageInput"
            name="Image"
            type="file"
          />
        </div>
        <hr />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="container p-1 d-flex flex-column align-items-center"
    >
      <h1
        class="p-2"
      >
        Add a class
      </h1>
      <input
        class="form-control rounded p-1 px-2"
        id="title"
        name="Title"
        placeholder="Enter a title"
      />
      <div
        class="invalid"
      />
      <textarea
        class="form-control rounded p-1 px-2"
        id="description"
        name="Description"
        placeholder="Enter a description"
      />
      <div
        class="invalid"
      />
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formFile"
        >
          Upload a class thumbnail
        </label>
        <input
          accept=".jpg, .png"
          class="form-control form-control-sm"
          data-testid="file-input"
          id="ImageInput"
          name="Image"
          type="file"
        />
      </div>
      <hr />
      <button
        class="btn btn-primary"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
